src/lib/metamaskInpageProvider.ts

```diff
@@ -0,0 +1,103 @@
+/// <reference path="./metamaskInpageProvider.d.ts" />
+
+import { initProvider } from 'metamask-inpage-provider'
+import PostMessageStream from 'post-message-stream'
+
+function universalProxy(pseudoTarget: object) {
+  return {
+    proxy: new Proxy(
+      {},
+      new Proxy(
+        {},
+        {
+          get(_, p) {
+            return (_t: any, p2: any, r: any) => {
+              switch (p) {
+                case 'get': {
+                  const out = Reflect.get(pseudoTarget, p2, r)
+                  if (typeof out === 'function') return out.bind(pseudoTarget)
+                  return out
+                }
+                case 'getOwnPropertyDescriptor': {
+                  const out = Reflect.getOwnPropertyDescriptor(pseudoTarget, p2)
+                  if (out) out.configurable = true
+                  return out
+                }
+                case 'isExtensible':
+                  return true
+                case 'preventExtensions':
+                  return false
+                default:
+                  return (Reflect as any)[p](pseudoTarget, p2, r)
+              }
+            }
+          }
+        }
+      )
+    ),
+    getPseudoTarget() {
+      return pseudoTarget
+    },
+    setPseudoTarget(value: object) {
+      pseudoTarget = value
+    }
+  }
+}
```

Comment:
This is built on the `universalProxyHandler` code in hdwallet's `revocable()`, so it should be pretty robust; a similar wrappers are around everything returned from a native vault. The difference here is that the (pseudo-)target can be changed after creation, which lets us return a "placeholder" proxy object for any on-load references to `window.ethereum`, which can be filled in later once we decide which stub to use.